flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -338,6 +338,9 @@ protected StreamTask(
                 Executors.newCachedThreadPool(
                         new ExecutorThreadFactory("AsyncOperations", uncaughtExceptionHandler));
 
+        environment.setMainMailboxExecutor(mainMailboxExecutor);
+        environment.setAsyncOperationsThreadPool(asyncOperationsThreadPool);
```

Comment:
Why isn't IO executor used here?

From its name, `asyncOperationsThreadPool` can be used for anything including "control" actions (like declining checkpoints) or collecting metrics - not only IO.

Though currently it is used for IO only - and that's why we should probably merge the two, likely in favor of Tasks IO executor.

This will also allow to inject it via constructor and make the field immutable.

WDYT?

cc: @pnowojski

Reply:
asyncOperationsThreadPool provides async snapshot workers for async checkpoint runnable, which is exactly what we used for materialization in the changelog case.

whether/how to merge IO pool with this async pool is a different topic, which is related but may not be within this ticket's scope.

Reply:
> asyncOperationsThreadPool provides async snapshot workers for async checkpoint runnable, which is exactly what we used for materialization in the changelog case.

I think it's not exactly true - `asyncOperationsThreadPool` only waits for snapshot completion, but upload can happen in other threads (which is the with RocksDB).

Merging these pools is out of scope (if it happens), but should be taken into account IMO.

Reply:
> I think it's not exactly true - asyncOperationsThreadPool only waits for snapshot completion, but upload can happen in other threads (which is the with RocksDB).

That's exactly happening here: `asyncOperationsThreadPool` is waiting for a snapshot completion. The difference is that for the normal state backend, async happens during checkpointing; while in materialization(changelog), it happens during periodic materialization.

> Merging these pools is out of scope (if it happens), but should be taken into account IMO.
That part I totally agree, that why `asyncOperationsThreadPool` is reused here.